src/builder.rs

```diff
@@ -37,8 +37,48 @@ pub struct RootedPath {
     pub path: StorePath,
 }
 
+/// A path that was accessed during instantiation and should be watched
+/// for changes.
+#[derive(Debug)]
+pub enum ReferencedPath {
+    /// Should watch the whole directory (if the path is a directory).
+    Recursive(PathBuf),
+    /// Just watch path that was accessed.
+    NotRecursive(PathBuf),
+}
```

Comment:
Would it not be simpler to track recursively and non-recursively watched paths in separate `Vec<PathBuf>`s?